ztp/siteconfig-generator/siteConfig/siteConfigBuilder_test.go

```diff
@@ -0,0 +1,160 @@
+	scb := SiteConfigBuilder{}
+	for _, test := range tests {
+		fn, content, _ := scb.getManifestFromTemplate(test.template, "role", test.data)
+		assert.Equal(t, test.expectFn, fn)
+		assert.Equal(t, test.expectContent, content)
+	}
+}
+
+func Test_siteConfigBuildValidation(t *testing.T) {
+
+	sc := SiteConfig{}
+	err := yaml.Unmarshal([]byte(siteConfigTest), &sc)
+	assert.Equal(t, err, nil)
+
+	scBuilder, _ := NewSiteConfigBuilder()
+	// Set empty cluster Name
+	sc.Spec.Clusters[0].ClusterName = ""
+	_, err = scBuilder.Build(sc)
+	assert.Equal(t, err, errors.New("Error: Missing cluster name at site test-site"))
+
+	// Set invalid network type
+	sc.Spec.Clusters[0].ClusterName = "cluster1"
+	sc.Spec.Clusters[0].NetworkType = "invalidNetworkType"
+	_, err = scBuilder.Build(sc)
+	assert.Equal(t, err, errors.New("Error: networkType must be either OpenShiftSDN or OVNKubernetes test-site/cluster1"))
+
+	// Set repeated cluster names
+	sc.Spec.Clusters[0].NetworkType = "OVNKubernetes"
+	sc.Spec.Clusters = append(sc.Spec.Clusters, sc.Spec.Clusters[0])
+	scBuilder.setLocalExtraManifestPath("../../source-crs/extra-manifest")
+	_, err = scBuilder.Build(sc)
+	assert.Equal(t, err, errors.New("Error: Repeated Cluster Name test-site/cluster1"))
+}
+
+func Test_siteConfigBuildExtraManifest(t *testing.T) {
+	sc := SiteConfig{}
+	err := yaml.Unmarshal([]byte(siteConfigTest), &sc)
+	assert.Equal(t, err, nil)
+
+	scBuilder, _ := NewSiteConfigBuilder()
+
+	// Expect to fail as the localExtraManifest path is in its default value
+	_, err = scBuilder.Build(sc)
+	assert.NotEqual(t, err, nil)
+	assert.Equal(t, strings.Contains(err.Error(), "no such file or directory"), true)
+
+	// Set the local extra manifest path to cnf-feature-deploy/ztp/source-crs/extra-manifest
+	scBuilder.setLocalExtraManifestPath("../../source-crs/extra-manifest")
+	// Setting the networkType to its default value
+	sc.Spec.Clusters[0].NetworkType = "OVNKubernetes"
+	clustersCRs, err := scBuilder.Build(sc)
+	assert.Equal(t, err, nil)
+	assert.NotEqual(t, clustersCRs["test-site/cluster1"], nil)
+	// expect all the installation CRs generated
+	assert.Equal(t, len(clustersCRs["test-site/cluster1"]), len(scBuilder.SourceClusterCRs))
+
+	// Setting invalid user extra manifest path
+	sc.Spec.Clusters[0].ExtraManifestPath = "invalid-path/extra-manifest"
+	sc.Spec.Clusters[0].NetworkType = "OVNKubernetes"
+	_, err = scBuilder.Build(sc)
+	assert.NotEqual(t, err, nil)
+	assert.Equal(t, strings.Contains(err.Error(), "no such file or directory"), true)
+
+	// Setting valid user extra manifest path
+	sc.Spec.Clusters[0].ExtraManifestPath = "testdata/user-extra-manifest"
+	sc.Spec.Clusters[0].NetworkType = "OVNKubernetes"
+	clustersCRs, err = scBuilder.Build(sc)
+	assert.Equal(t, err, nil)
+	assert.NotEqual(t, clustersCRs["test-site/cluster1"], nil)
+	// expect all the installation CRs generated
+	assert.Equal(t, len(clustersCRs["test-site/cluster1"]), len(scBuilder.SourceClusterCRs))
```

Comment:
This is a good start but doesn't test the most important part of the extramanifest logic that changed:
- An additional file from the ExtraManifestPath with a unique name is added to the ConfigMap
- An additional file from the ExtraManifestPath with a name collision with a default manifest is not allowed

Reply:
Note: you could have a separate unit test around `getExtraManifest` specifically, it doesn't need to use `Build` as the entrypoint for all unit tests.

Reply:
ops, forget to add the check for preventing override. will add